File: Java/src/com/dbc/algorithm/Daily/NumberOfLongestIncreasingSubsequence.java
package com.dbc.algorithm.Daily;

import java.util.Arrays;

public class NumberOfLongestIncreasingSubsequence {
    public static int findNumberOfLISDp(int[] nums) {
        if (nums.length <= 1) {
            return nums.length;
        }

        int[] lengths = new int[nums.length];
        int[] counts = new int[nums.length];
        Arrays.fill(lengths, 0);
        Arrays.fill(counts, 1);

        for (int j = 0; j < nums.length; j++) {
            for (int i = 0; i < j; i++) {
                if (nums[i] < nums[j]) {
                    if (lengths[i] >= lengths[j]) {
                        lengths[j] = 1 + lengths[i];
                        counts[j] = counts[i];
                    } else if (lengths[i] + 1 == lengths[j]) {
                        counts[j] += counts[i];
                    }
                }
            }
        }

        int lengthiest = 0;
        for (int i = 0; i < nums.length; i++) {
            lengthiest = Math.max(lengthiest, lengths[i]);
        }

        int result = 0;
        for (int i = 0; i < nums.length; i++) {
            if (lengthiest == lengths[i])
                result += counts[i];
        }

        return result;
    }

    public Value merge(Value v1, Value v2) {
        if (v1.length == v2.length) {
            if (v1.length == 0) return new Value(0, 1);
            return new Value(v1.length, v1.count + v2.count);
        }
        return v1.length > v2.length ? v1 : v2;
    }

    public void insert(Node node, int key, Value val) {
        if (node.range_left == node.range_right) {
            node.val = merge(val, node.val);
            return;
        } else if (key <= node.getRangeMid()) {
            insert(node.getLeft(), key, val);
        } else {
            insert(node.getRight(), key, val);
        }
        node.val = merge(node.getLeft().val, node.getRight().val);
    }

    public Value query(Node node, int key) {
        if (node.range_right <= key) return node.val;
        else if (node.range_left > key) return new Value(0, 1);
        else return merge(query(node.getLeft(), key), query(node.getRight(), key));
    }

    public int findNumberOfLIS_tree(int[] nums) {
        if (nums.length == 0) return 0;
        int min = nums[0], max = nums[0];
        for (int num: nums) {
            min = Math.min(min, num);
            max = Math.max(max, num);
        }
        Node root = new Node(min, max);
        for (int num: nums) {
            Value v = query(root, num-1);
            insert(root, num, new Value(v.length + 1, v.count));
        }
        return root.val.count;
    }

    class Node {
        int range_left, range_right;
        Node left, right;
        Value val;
        public Node(int start, int end) {
            range_left = start;
            range_right = end;
            left = null;
            right = null;
            val = new Value(0, 1);
        }
        public int getRangeMid() {
            return range_left + (range_right - range_left) / 2;
        }
        public Node getLeft() {
            if (left == null) left = new Node(range_left, getRangeMid());
            return left;
        }
        public Node getRight() {
            if (right == null) right = new Node(getRangeMid() + 1, range_right);
            return right;
        }
    }

    class Value {
        int length;
        int count;
        public Value(int len, int ct) {
            length = len;
            count = ct;
        }
    }
}

File: Python/dynamic-programming/number_of_longest_increasing_subsequence.py
def findNumberOfLIS_dp(nums: list) -> int:
    length = len(nums)
    if length <= 1:
        return length

    lengths = [0] * length
    counts = [1] * length

    for j, num in enumerate(nums):
        for i in range(j):
            if nums[i] < nums[j]:
                if lengths[i] >= lengths[j]:
                    lengths[j] = 1 + lengths[i]
                    counts[j] = counts[i]
                elif lengths[i] + 1 == lengths[j]:
                    counts[j] += counts[i]

    longest = max(lengths)
    return sum(c for i, c in enumerate(counts) if lengths[i] == longest)


class Node(object):
    def __init__(self, start, end):
        self.range_left, self.range_right = start, end
        self._left = self._right = None
        self.val = 0, 1 #length, count
    @property
    def range_mid(self):
        return (self.range_left + self.range_right) / 2
    @property
    def left(self):
        self._left = self._left or Node(self.range_left, self.range_mid)
        return self._left
    @property
    def right(self):
        self._right = self._right or Node(self.range_mid + 1, self.range_right)
        return self._right

def merge(v1, v2):
    if v1[0] == v2[0]:
        if v1[0] == 0: return (0, 1)
        return v1[0], v1[1] + v2[1]
    return max(v1, v2)

def insert(node, key, val):
    if node.range_left == node.range_right:
        node.val = merge(val, node.val)
        return
    if key <= node.range_mid:
        insert(node.left, key, val)
    else:
        insert(node.right, key, val)
    node.val = merge(node.left.val, node.right.val)

def query(node, key):
    if node.range_right <= key:
        return node.val
    elif node.range_left > key:
        return 0, 1
    else:
        return merge(query(node.left, key), query(node.right, key))


def findNumberOfLIS_tree(nums: list) -> int:
    if not nums: return 0
    root = Node(min(nums), max(nums))
    for num in nums:
        length, count = query(root, num-1)
        insert(root, num, (length+1, count))
    return root.val[1]


if __name__ == "__main__":
    print(findNumberOfLIS_tree([2, 2, 2, 2, 2]))
